libsupport/include/katana/Experimental.h

```diff
@@ -0,0 +1,96 @@
+#ifndef KATANA_LIBSUPPORT_KATANA_EXPERIMENTAL_H_
+#define KATANA_LIBSUPPORT_KATANA_EXPERIMENTAL_H_
+
+#include <memory>
+#include <mutex>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "katana/config.h"
+
+namespace katana {
+
+/// ExperimentalFeature tracks the state of feature flags set in the environment;
+/// It is not intended to be used directly; please see the macro KATANA_EXPERIMENTAL_FEATURE
+/// below
+class KATANA_EXPORT ExperimentalFeature {
+public:
+  static ExperimentalFeature* Register(const std::string& feature_name);
+
+  /// report the feature flags that were checked on codepaths that were executed and
+  /// the flag was set to true
+  static std::vector<std::string> ReportEnabled();
+
+  /// report the feature flags that were used but stayed false
+  static std::vector<std::string> ReportDisabled();
+
+  /// report the feature flags that were provided but did not match any registered flag
+  static std::vector<std::string> ReportUnrecognized();
+
+  bool IsEnabled() { return is_enabled_; }
+
+  ExperimentalFeature(const ExperimentalFeature& no_copy) = delete;
+  ExperimentalFeature& operator=(const ExperimentalFeature& no_copy) = delete;
+  ExperimentalFeature(ExperimentalFeature&& no_move) = delete;
+  ExperimentalFeature& operator=(ExperimentalFeature&& no_move) = delete;
+  ~ExperimentalFeature() = default;
+
+private:
+  ExperimentalFeature(std::string name) : name_(std::move(name)) { CheckEnv(); }
+
+  void CheckEnv();
+
+  std::string name_;
+  bool is_enabled_;
+
+  static std::mutex registered_features_mutex_;
+  static std::unordered_map<
+      std::string, std::unique_ptr<katana::ExperimentalFeature>>
+      registered_features_;
+};
+
+}  // namespace katana
+
+/// KATANA_EXPERIMENTAL_FEATURE creates a flag that can be set from the environment.
+/// The macro takes a feature_name which should be an unquoted, unique string that
+/// looks like a function name. It is expanded to the function:
+/// bool Feature_<feature_name>_IsEnabled(). Developers can then call this function
+/// to detect if the flag was set.
+///
+/// Flags are set using the environment variable KATANA_ENABLE_EXPERIMENTAL. Users
+/// pass the same string passed to KATANA_EXPERIMENTAL_FEATURE to set a particular
+/// flag. Multiple flags may be set by providing a comma delimited list of feature
+/// names.
+///
+/// Example:
+///
+/// Program env:
+///   KATANA_ENABLE_EXPERIMENTAL="UnstableButUseful"
+///
+/// active_development.cpp:
+///    KATANA_EXPERIMENTAL_FEATURE(UnstableButUseful);
+///
+///    void important_function() {
```

Comment:
We're going with the assumption that classes/types, function declarations, or other compile-time changes are not going to depend on the experimental features. That's fine with me. It's good to state that up-front.

Reply:
Will add that. If we want compile time experimental features like that I think it should be a separate mechanism.

Reply:
Agreed. We should try to stick to environmental variable approach as much as possible because that allows us to try these features on customer deployments.